keps/sig-storage/1682-csi-driver-skip-permission/README.md

```diff
@@ -0,0 +1,227 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+
+## Summary
+
+Currently before bind mounting a CSI volume for a Pod, we use imprecise heuristics,
+such as presence of fsType on the PVC to determine if the volume supports fsGroup based
+permission change. These heuristics are known to be fragile, and cause problems with different
+storage types.
+
+To solve this issue we will add a new field called `CSIDriver.Spec.SupportsFSGroup` 
+that allows the driver to define if it supports volume ownership modifications via
+fsGroup.
+
+## Motivation
+
+When a volume is mounted on the node, we modify volume ownership 
+before bind mounting the volume inside container; however, if the volume
+does not support these operations (such as NFS), then the change is still attempted. 
+This results in errors being reported to the user if the volume doesn't 
+support these operations.
+
+### Goals
+
+ - Allow CSIDrivers to opt-in to volume ownership changes.
+
+### Non-Goals
+
+ - This does not attempt to update existing CSIDrivers themselves.
+
+## Proposal
+
+We propose that the `CSIDriver` type include a field that defines if the volume 
+provided by the driver supports changing volume ownership. This will be enabled
+with a new feature gate, `CSIVolumeSupportFSGroup`.   
+
+### Risks and Mitigations
+
+- CSIDrivers will need to be redeployed after this field is introduced if the desired behavior is modified.
+- If a cluster enables the `CSIVolumeSupportFSGroup` feature gate and then this feature gate is disabled,
+such as due to an upgrade or downgrade, then the cluster will revert to the current behavior of examining
+volumes and attempting to apply volume ownerships and permissions based on the defined `fsGroup`.
+
+## Design Details
+
+Currently Volume permission and ownership change is done automatically for every volume. 
+As part of this proposal we will change the algorithm that modifies volume ownership and permissions
+for CSIDrivers to check the new field, and skip volume ownership modifications if it is found to be
+`false`.
+
+When defining a `CSIDriver`, we propose that `CSIDriver.Spec` be expanded to include a new field entitled 
+`SupportsFSGroup` which can have following possible values:
+
+ - `OnlyRWO` --> Current behavior. Attempt to modify the volume ownership and permissions to the defined `fsGroup` when the volume is 
+ mounted if accessModes is RWO.
+ - `Never` --> New behavior. Attach the volume without attempting to modify volume ownership or permissions.
+ - `Always` --> New behavior. Always attempt to apply the defined fsGroup to modify volume ownership and permissions.
+
+```go
+type CSIDriverFSGroupChangePolicy string
```

Comment:
We can debate about naming later, but this may be confusing with the FSGroupChangePolicy in the PodSecurityContext

Reply:
What about something like `CSIDriverSupportsVolumeOwnershipChange` ? This would be a little clearer with the intent as well, although it is wordy.

Reply:
Maybe just `SupportsFsGroup`? Then it also aligns with the selinux proposal.

Reply:
I've updated this to be `SupportsFsGroup`. Should this be `SupportFSGroup` (two capital letters in "FS") ?